rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala

```diff
@@ -0,0 +1,218 @@
+    val key = hashChannels(channels)
+    for {
+      storedLeaf <- fetchData(key)
+      result = storedLeaf match {
+        case Some(ContinuationsLeaf(bytes)) =>
+          codecSeqWaitingContinuation.decode(bytes.bits).get.value
+        case _ => Seq.empty
+      }
+    } yield result
+  }
+
+  type Result = (Blake2b256Hash, Option[PersistedData], HistoryAction)
+
+  private def transform(actions: List[HotStoreAction]): List[Result] =
+    actions.map {
+      case i: InsertData[C, A] =>
+        val key      = hashChannel(i.channel)
+        val data     = codecSeqDatum.encode(i.data).get.toByteVector
+        val dataLeaf = DataLeaf(data)
+        val dataHash = Blake2b256Hash.create(data)
+        (dataHash, Some(dataLeaf), InsertAction(key.bytes.toSeq.toList, dataHash))
+      case i: InsertContinuations[C, P, K] =>
+        val key               = hashChannels(i.channels)
+        val data              = codecSeqWaitingContinuation.encode(i.continuations).get.toByteVector
+        val continuationsLeaf = ContinuationsLeaf(data)
+        val continuationsHash = Blake2b256Hash.create(data)
+        (
+          continuationsHash,
+          Some(continuationsLeaf),
+          InsertAction(key.bytes.toSeq.toList, continuationsHash)
+        )
+      case i: InsertJoins[C] =>
+        val key       = hashJoinsChannel(i.channel)
+        val data      = codecSeqSeqC.encode(i.joins).get.toByteVector
+        val joinsLeaf = JoinsLeaf(data)
+        val joinsHash = Blake2b256Hash.create(data)
+        (joinsHash, Some(joinsLeaf), InsertAction(key.bytes.toSeq.toList, joinsHash))
+      case d: DeleteData[C] =>
+        val key = hashChannel(d.channel)
+        (key, None, DeleteAction(key.bytes.toSeq.toList))
+      case d: DeleteContinuations[C] =>
+        val key = hashChannels(d.channels)
+        (key, None, DeleteAction(key.bytes.toSeq.toList))
+      case d: DeleteJoins[C] =>
+        val key = hashJoinsChannel(d.channel)
+        (key, None, DeleteAction(key.bytes.toSeq.toList))
+    }
+
+  private def storeLeaves(leafs: List[Result]): F[List[HistoryAction]] =
+    leafs.traverse {
+      case (key, Some(data), historyAction) =>
+        leafStore.put(key, data).map(_ => historyAction)
+      case (_, None, historyAction) =>
+        Applicative[F].pure(historyAction)
+    }
+
+  override def process(actions: List[HotStoreAction]): F[HistoryRepository[F, C, P, A, K]] =
+    for {
+      trieActions    <- Applicative[F].pure(transform(actions))
+      historyActions <- storeLeaves(trieActions)
+      nextHistory    <- history.process(historyActions)
+    } yield this.copy(history = nextHistory)
+}
+
+object HistoryRepositoryImpl {
+  implicit def codecSeq[A](implicit codecA: Codec[A]): Codec[Seq[A]] =
+    seqOfN(int32, codecA)
+
+  /**
+    * Codec that encodes/decodes a `Seq[A]` of `N` elements using a `Codec[A]`.
+    */
+  def seqOfN[A](countCodec: Codec[Int], valueCodec: Codec[A]): Codec[Seq[A]] =
```

Comment:
After my recent change you should be able to re-use `coop.rchain.scodec.codecs.seqOfN`

Reply:
yeah, will cut this asap